### PR TITLE
feat: introduce basic extended addresses implementation, stop using `platform{HTTP,P2P}Port`

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Depends on https://github.com/dashpay/dash/pull/6665

## Breaking Changes

_Work in progress._

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/rpc/masternode.cpp [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include <masternode/payments.h>
 #include <net.h>
 #include <netbase.h>
+#include <rpc/evo_util.h>
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/rpc/masternode.cpp:17]: Resolved in latest push

File: test/functional/rpc_netinfo.py [hunk 2]
```diff
@@ -202,15 +236,125 @@ def test_validation_legacy(self):
...
+        self.node_evo.register_mn(self, False, f"127.0.0.1:{self.node_evo.mn.nodePort}", "", DEFAULT_PORT_PLATFORM_HTTP,
+                                  -8, "Cannot leave platformP2PAddrs empty if other address fields populated")
+        self.node_evo.register_mn(self, False, "", DEFAULT_PORT_PLATFORM_P2P, "",
+                                  -8, "Must specify coreP2PAddrs in order to set platformP2PAddrs")
+        self.node_evo.register_mn(self, False, f"127.0.0.1:{self.node_evo.mn.nodePort}", DEFAULT_PORT_PLATFORM_P2P, "",
+                                  -8, "Cannot leave platformHTTPSAddrs empty if other address fields populated")
+
+        # coreP2PAddrs cannot omit the port, extended addresses require explicit port specification
+        self.node_evo.register_mn(self, False, f"127.0.0.1", DEFAULT_PORT_PLATFORM_P2P, DEFAULT_PORT_PLATFORM_HTTP,
+                                  -8, "Error setting coreP2PAddrs[0] to '127.0.0.1' (invalid port)")
```
[the remark this answers is not in the document]
Comment (kwvg) [reply test/functional/rpc_netinfo.py]: Resolved in latest push

File: test/functional/rpc_netinfo.py [hunk 3]
```diff
@@ -57,68 +66,54 @@ def is_mn_visible(self, _protx_hash = None) -> bool:
...
-    def register_mn(self, test: BitcoinTestFramework, submit: bool, addrs_core_p2p, addrs_platform_p2p = None, addrs_platform_http = None, code = None, msg = None) -> str:
-        assert self.mn.nodeIdx is not None
+    def set_active_state(self, test: BitcoinTestFramework, active: bool, more_extra_args: Optional[List[str]] = None):
+        assert self.mn.proTxHash and self.mn.keyOperator
+        target_extra_args = self.node.extra_args.copy()
+        if active:
+            target_extra_args += [f'-masternodeblsprivkey={self.mn.keyOperator}']
+        if more_extra_args is not None:
+            target_extra_args += more_extra_args
+        test.restart_node(self.mn.nodeIdx, extra_args=target_extra_args)
```
[the remark this answers is not in the document]
Comment (kwvg) [reply test/functional/rpc_netinfo.py:81]: We're copying the original list of arguments from node startup, not the amended list from each restart, we are keeping a `copy()` to avoid mutating the original list (which will prevent using RPCs needed by the test), existing code works as-is.

File: src/evo/netinfo.cpp [hunk 4]
```diff
@@ -244,13 +244,18 @@ NetInfoStatus MnNetInfo::Validate() const
 
 UniValue MnNetInfo::ToJson() const
 {
+    if (IsEmpty()) {
+        return UniValue(UniValue::VARR);
```
Comment (knst) [inline src/evo/netinfo.cpp]: nit: `UniValue{UniValue::VARR};` is better here I think

File: src/test/evo_netinfo_tests.cpp [hunk 5]
```diff
@@ -126,6 +126,16 @@ BOOST_FIXTURE_TEST_CASE(extnetinfo_rules_reg, RegTestingSetup)
...
     }
+
+    {
+        // ExtNetInfo does not allow storing duplicates
+        ExtNetInfo netInfo;
+        BOOST_CHECK_EQUAL(netInfo.AddEntry("1.1.1.1:9998"), NetInfoStatus::Success);
+        // Exact duplicates are prohibited
+        BOOST_CHECK_EQUAL(netInfo.AddEntry("1.1.1.1:9998"), NetInfoStatus::Duplicate);
+        // Partial duplicates (same address, different port) are also prohibited
+        BOOST_CHECK_EQUAL(netInfo.AddEntry("1.1.1.1:9997"), NetInfoStatus::Duplicate);
```
Comment (knst) [inline src/test/evo_netinfo_tests.cpp]: nit: may you add tests with duplicates of ipv6 address too?

File: src/evo/netinfo.cpp [hunk 6]
```diff
@@ -311,8 +323,12 @@ NetInfoStatus ExtNetInfo::ValidateService(const CService& service)
     return NetInfoStatus::Success;
 }
 
-NetInfoStatus ExtNetInfo::AddEntry(const std::string& input)
+NetInfoStatus ExtNetInfo::AddEntry(const uint8_t purpose, const std::string& input)
```
Comment (knst) [inline src/evo/netinfo.cpp]: nit: purpose here should have a type of enum, not `uint8_t`

File: src/evo/netinfo.h [hunk 7]
```diff
@@ -60,6 +60,35 @@ constexpr std::string_view NISToString(const NetInfoStatus code)
     assert(false);
 }
 
+// Purpose corresponds to the index position in the ExtNetInfo map (which is ordered),
+// entries must be contiguous and cannot be changed once set without a format version
+// update, which will necessitate a hard-fork.
+namespace Purpose {
+enum : uint8_t {
```
Comment (knst) [inline src/evo/netinfo.h]: nit: `enum class` for all new code is preferable.

File: src/evo/netinfo.h [hunk 8]
```diff
@@ -60,6 +60,35 @@ constexpr std::string_view NISToString(const NetInfoStatus code)
...
+{
+    switch (purpose) {
+    case Purpose::CORE_P2P:
+        return true;
+    } // no default case, so the compiler can warn about missing cases
+    return false;
+}
+
+// Warning: Used in RPC code, altering existing values is a breaking change
+constexpr std::string_view PurposeToString(const uint8_t purpose, const bool lower = false)
```
Comment (knst) [inline src/evo/netinfo.h]: here should be enum instead uint8_t

```
constexpr std::string_view PurposeToString(const Purpose purpose, const bool lower = false)

File: src/evo/netinfo.h [hunk 9]
```diff
@@ -60,6 +60,35 @@ constexpr std::string_view NISToString(const NetInfoStatus code)
     assert(false);
 }
 
+// Purpose corresponds to the index position in the ExtNetInfo map (which is ordered),
+// entries must be contiguous and cannot be changed once set without a format version
+// update, which will necessitate a hard-fork.
+namespace Purpose {
```
Comment (knst) [inline src/evo/netinfo.h]: nit: this name is way to general for using in global namespace; consider using here instead of just Purpose something like NetPurpose or `AddressPurpose` or NetInfoPurpose` or something other less generic

File: src/evo/netinfo.h [hunk 10]
```diff
@@ -146,11 +175,12 @@ class NetInfoInterface
 public:
     virtual ~NetInfoInterface() = default;
 
-    virtual NetInfoStatus AddEntry(const std::string& service) = 0;
+    virtual NetInfoStatus AddEntry(const uint8_t purpose, const std::string& service) = 0;
     virtual NetInfoList GetEntries() const = 0;
 
     virtual CService GetPrimary() const = 0;
     virtual bool CanStorePlatform() const = 0;
+    virtual bool HasEntries(uint8_t purpose) const = 0;
```
Comment (knst) [inline src/evo/netinfo.h]: consider passing here enum instead uint8_t

File: src/evo/netinfo.h [hunk 11]
```diff
@@ -203,10 +233,11 @@ class MnNetInfo final : public NetInfoInterface
         m_addr = NetInfoEntry{service};
     }
 
-    NetInfoStatus AddEntry(const std::string& service) override;
+    NetInfoStatus AddEntry(const uint8_t purpose, const std::string& service) override;
```
Comment (knst) [inline src/evo/netinfo.h]: enum too

File: src/evo/netinfo.h [hunk 12]
```diff
@@ -203,10 +233,11 @@ class MnNetInfo final : public NetInfoInterface
...
     NetInfoList GetEntries() const override;
 
     CService GetPrimary() const override;
+    bool HasEntries(uint8_t purpose) const override { return purpose == Purpose::CORE_P2P && !IsEmpty(); }
```
Comment (knst) [inline src/evo/netinfo.h]: enum too

File: src/test/evo_netinfo_tests.cpp [hunk 13]
```diff
@@ -15,33 +15,35 @@
 
 BOOST_FIXTURE_TEST_SUITE(evo_netinfo_tests, BasicTestingSetup)
 
-using TestVectors =
-    std::vector<std::tuple</*input=*/std::string, /*expected_ret_mn=*/NetInfoStatus, /*expected_ret_ext=*/NetInfoStatus>>;
+using TestVectors = std::vector<std::tuple</*input=*/std::pair</*purpose=*/uint8_t, /*addr=*/std::string>,
```
Comment (knst) [inline src/test/evo_netinfo_tests.cpp]: make it an enum here, so, you don't need to write a doc `/*purpose=*/` then

File: src/evo/netinfo.h (hunk 8 above)
Comment (knst) [reply src/evo/netinfo.h]: it seems as capitals are used only in logs. If it is true, consider to drop an argument `lower=false` and just always output it (in RPC and in logs) as lowercase without capital option. It doesn't seems much useful for me

File: src/rpc/coinjoin.cpp [hunk 14]
```diff
@@ -440,6 +440,16 @@ static RPCHelpMan getcoinjoininfo()
                                                 {RPCResult::Type::STR, "address", ""},
                                             }
                                         },
+                                        {RPCResult::Type::ARR, "platform_p2p", /*optional=*/true, "Addresses used for Platform P2P",
```
Comment (knst) [inline src/rpc/coinjoin.cpp]: nit: if you will keep it here, consider mentioning evo nodes here; 
maybe something like
```
                                        {RPCResult::Type::ARR, "platform_p2p", /*optional=*/true, "Addresses used for Platform P2P (EvoNodes only)",
```

Comment (knst) [inline src/rpc/coinjoin.cpp]: IMO coinjoin related RPC `getcoinjoininfo` should not return `platform_https` and `platform_p2p`;  consider just removing these fields comletely.
Probably you need to add helper "get p2p addreses only" for NetInfo, something like that:
```
-        obj.pushKV("addresses", mixingMasternode->pdmnState->netInfo->ToJson());
+        obj.pushKV("addresses", mixingMasternode->pdmnState->netInfo->FilterBy(NetInfoPurpose::P2P_CORE)->ToJson());
```

File: src/evo/deterministicmns.cpp [hunk 15]
```diff
@@ -901,39 +901,43 @@ static bool CheckService(const ProTx& proTx, TxValidationState& state)
...
         }
+        if (proTx.platformHTTPPort != 0) {
+            return state.Invalid(TxValidationResult::TX_BAD_SPECIAL, "bad-protx-platform-http-port");
+        }
+        return true;
     }
 
     // TODO: use real args here
-    static int mainnetPlatformHTTPPort = CreateChainParams(ArgsManager{}, CBaseChainParams::MAIN)->GetDefaultPlatformHTTPPort();
+    const auto main_params{CreateChainParams(ArgsManager{}, CBaseChainParams::MAIN)};
```
Comment (knst) [inline src/evo/deterministicmns.cpp]: wait, why here `CreateChainParams()`?
I haven't noticed it before in prior PRs... Why can't you just use here `Params()` from header `chainparams.h`?


Calling `CreateChainParams` is not good at least from performance point of view. It's not a very heavy object, but it's not cheap at all

File: src/evo/providertx.cpp [hunk 16]
```diff
@@ -187,10 +190,13 @@ std::string CProUpServTx::ToString() const
     }
 
     return strprintf("CProUpServTx(nVersion=%d, nType=%d, proTxHash=%s, operatorPayoutAddress=%s, "
-                     "platformNodeID=%s, platformP2PPort=%d, platformHTTPPort=%d)\n"
+                     "platformNodeID=%s%s)\n"
                      "  %s",
```
Comment (knst) [inline src/evo/providertx.cpp]: nit: consider merging 2 lines for simplicity

```diff
-                     "platformNodeID=%s%s)\n"
-                     "  %s",
+                     "platformNodeID=%s%s)\n   %s",
```

File: src/rpc/evo.cpp [hunk 17]
```diff
@@ -184,12 +184,14 @@ static RPCArg GetRpcArg(const std::string& strParamName)
                 "Platform P2P node ID, derived from P2P public key."}
         },
         {"platformP2PPort",
-            {"platformP2PPort", RPCArg::Type::NUM, RPCArg::Optional::NO,
-                "TCP port of Dash Platform peer-to-peer communication between nodes (network byte order)."}
+            {"platformP2PPort", RPCArg::Type::STR, RPCArg::Optional::NO,
```
Comment (knst) [inline src/rpc/evo.cpp]: it's not more `platormP2PPort`; especially if it's string.

I'd suggest to create new field if you really need to return string for any reason; otherwise just return PORT.

Also I see further you requires platform address to has port only (not address), so, I'd suggest to don't change this one.
See:
```
                                  -8, "ProTx version disallows storing addresses in platformP2PPort (must specify port number only)")
```

**Review by knst — COMMENTED** [review]
This PR is really big! I reviewed half of commits, please consider my comments

File: src/rpc/evo.cpp [hunk 18]
```diff
@@ -184,24 +184,36 @@ static RPCArg GetRpcArg(const std::string& strParamName)
...
+                {
+                    {"address", RPCArg::Type::STR, RPCArg::Optional::NO, ""},
+                }}
         },
         {"platformHTTPPort",
-            {"platformHTTPPort", RPCArg::Type::STR, RPCArg::Optional::NO,
-                "Address in the form \"ADDR:PORT\" used by Platform for their HTTPS API.\n"
-                "Must be unique on the network. Can be set to an empty string, which will require a ProUpServTx afterwards."}
+            {"platformHTTPPort", RPCArg::Type::ARR, RPCArg::Optional::NO,
+                "Array of addresses in the form \"ADDR:PORT\" used by Platform for their HTTPS API.\n"
```
Comment (knst) [inline src/rpc/evo.cpp:204]: <s>
this one is strange again. here code says:

                "Array of addresses in the form \"ADDR:PORT\" used by Platform for their HTTPS API.\n"

but below JSONRPCError says: `ProTx version disallows storing addresses in platformP2PPort (must specify port number only)`

I can not understand it, probably end-point user won't understand it too.

Further refactoring `platformP2PPort` -> `platformP2PAddrs` make it even worse
</s>

UPD: error below is only for old version of protx, nevermind

File: src/evo/dmnstate.cpp [hunk 19]
```diff
@@ -131,5 +131,37 @@ UniValue CDeterministicMNStateDiff::ToJson(MnType nType) const
...
+            }
+            if (fields & Field_platformHTTPPort) {
+                netInfoObj.pushKV(PurposeToString(Purpose::PLATFORM_HTTPS, /*lower=*/true).data(),
+                                  (has_netinfo)
+                                      ? ArrFromService(CService(state.netInfo->GetPrimary(), state.platformHTTPPort))
+                                      : unknownAddr(state.platformHTTPPort));
+            }
+        }
+        if (!netInfoObj.empty()) {
+            obj.pushKV("addresses", netInfoObj);
```
Comment (knst) [inline src/evo/dmnstate.cpp:162]: nit: IMO should not re-order fields if there's no reason to do it.

`addresses` will be now the last fields

File: src/evo/netinfo.h [hunk 20]
```diff
@@ -246,7 +246,7 @@ class MnNetInfo final : public NetInfoInterface
     }
 
     NetInfoStatus AddEntry(const uint8_t purpose, const std::string& service) override;
-    NetInfoList GetEntries() const override;
+    NetInfoList GetEntries(std::optional<uint8_t> purpose_opt = std::nullopt) const override;
```
Comment (knst) [inline src/evo/netinfo.h]: std::optional here looks a bit too complicated; maybe better to add one more _fake_ value to enum such as `Purpose::ALL` ?

File: test/functional/test_framework/test_framework.py [hunk 21]
```diff
@@ -1242,7 +1242,7 @@ def register(self, node: TestNode, submit: bool, collateral_txid: Optional[str]
         args = [
             collateral_txid or self.collateral_txid,
             collateral_vout or self.collateral_vout,
-            addrs_core_p2p or [f'127.0.0.1:{self.nodePort}'],
+            [f'127.0.0.1:{self.nodePort}'] if addrs_core_p2p is None else addrs_core_p2p,
```
Comment (knst) [inline test/functional/test_framework/test_framework.py:1248]:
```python
            [f'127.0.0.1:{self.nodePort}'] if addrs_core_p2p is None else addrs_core_p2p,

            [f'127.0.0.1:{self.nodePort}'] if addrs_core_p2p is None else addrs_core_p2p,

            [f'127.0.0.1:{self.nodePort}'] if addrs_core_p2p is None else addrs_core_p2p,
```

nit: consider create a helper here

File: test/functional/rpc_netinfo.py [hunk 22]
```diff
@@ -126,39 +132,61 @@ def destroy_mn(self, test: BitcoinTestFramework):
...
+        self.num_nodes = 3
+        self.extra_args = [[
+            "-dip3params=2:2", f"-vbparams=v23:{self.mocktime}:999999999999:{V23_ACTIVATION_THRESHOLD}:10:8:6:5:0"
+        ] for _ in range(self.num_nodes)]
+        self.extra_args[2] += ["-deprecatedrpc=service"]
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
+    def activate_v23(self):
```
Comment (knst) [inline test/functional/rpc_netinfo.py:147]: nit: consider using here `activate_by_name` instead self-written helper

File: src/evo/netinfo.cpp (hunk 4 above)
Comment (kwvg) [reply src/evo/netinfo.cpp]: Resolved in latest push.

File: src/test/evo_netinfo_tests.cpp (hunk 5 above)
Comment (kwvg) [reply src/test/evo_netinfo_tests.cpp]: Any `CService`-representable address would follow the same code path

File: src/evo/netinfo.cpp (hunk 6 above)
Comment (kwvg) [reply src/evo/netinfo.cpp]: Changed argument type to `NetInfoPurpose` and used `enum class`, resolved in latest push.

File: src/evo/netinfo.h (hunk 7 above)
Comment (kwvg) [reply src/evo/netinfo.h]: See [comment](https://github.com/dashpay/dash/pull/6666#discussion_r2286472412), resolved in latest push.

File: src/evo/netinfo.h (hunk 8 above)
Comment (kwvg) [reply src/evo/netinfo.h]: Used lowercase and dropped `lower` argument, resolved in latest push.

File: src/evo/netinfo.h (hunk 9 above)
Comment (kwvg) [reply src/evo/netinfo.h]: Renamed to `NetInfoPurpose`, resolved in latest push

File: src/evo/netinfo.h (hunk 10 above)
Comment (kwvg) [reply src/evo/netinfo.h]: See [comment](https://github.com/dashpay/dash/pull/6666#discussion_r2286472412), resolved in latest push.

File: src/evo/netinfo.h (hunk 11 above)
Comment (kwvg) [reply src/evo/netinfo.h]: See [comment](https://github.com/dashpay/dash/pull/6666#discussion_r2286472412), resolved in latest push.

File: src/evo/netinfo.h (hunk 12 above)
Comment (kwvg) [reply src/evo/netinfo.h]: See [comment](https://github.com/dashpay/dash/pull/6666#discussion_r2286472412), resolved in latest push.

File: src/test/evo_netinfo_tests.cpp (hunk 13 above)
Comment (kwvg) [reply src/test/evo_netinfo_tests.cpp]: Replaced with struct `TestEntry`

File: src/rpc/coinjoin.cpp (hunk 14 above)
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: Resolved in latest push

File: src/evo/providertx.cpp (hunk 16 above)
Comment (kwvg) [reply src/evo/providertx.cpp]: Log format was changed to be single-line, realigned with the remaining set of printed values, resolved in latest push.

File: test/functional/rpc_netinfo.py (hunk 22 above)
Comment (kwvg) [reply test/functional/rpc_netinfo.py:147]: We don't use `DashTestFramework` in `rpc_netinfo.py` as we need to create and destroy nodes using unexpected arguments

File: src/rpc/coinjoin.cpp (hunk 14 above)
Comment (PastaPastaPasta) [reply src/rpc/coinjoin.cpp]: Agreed, although this wouldn't be a blocker

File: src/evo/core_write.cpp [hunk 23]
```diff
@@ -74,7 +75,7 @@
     if (IsServiceDeprecatedRPCEnabled()) {
         ret.pushKV("service", netInfo->GetPrimary().ToStringAddrPort());
     }
-    ret.pushKV("addresses", netInfo->ToJson());
+    ret.pushKV("addresses", ShimNetInfoPlatform(*this, nType));
```
Comment (PastaPastaPasta) [inline src/evo/core_write.cpp]: let's rename this, to make it clearer how it is intended to be used. This is a bad name, but `MakeNetInfoFromNetInfoOrLegacy` or something like that. @coderabbitai any ideas on a better name?

File: src/evo/netinfo.h [hunk 24]
```diff
@@ -219,6 +269,95 @@ class MnNetInfo final : public NetInfoInterface
...
+    //! Returns true if candidate is an addr:port duplicate in the object
+    bool IsAddrPortDuplicate(const NetInfoEntry& candidate) const;
+
+    //! Returns true if there are addr duplicates within a given address list
+    bool HasAddrDuplicates(const NetInfoList& entries) const;
+
+    //! Returns true if candidate is an addr duplicate within a given address list
+    bool IsAddrDuplicate(const NetInfoEntry& candidate, const NetInfoList& entries) const;
+
+    NetInfoStatus ProcessCandidate(const NetInfoPurpose purpose, const NetInfoEntry& candidate);
```
Comment (PastaPastaPasta) [inline src/evo/netinfo.h:296]: add docs

File: test/functional/test_framework/test_framework.py [hunk 25]
```diff
@@ -1415,27 +1410,18 @@ def update_registrar(self, node: TestNode, submit: bool, pubKeyOperator: Optiona
 
         return ret
 
-    def update_service(self, node: TestNode, submit: bool, coreP2PAddrs: Union[str, List[str], None] = None, platform_node_id: Optional[str] = None, platform_p2p_port: Optional[int] = None,
-                       platform_http_port: Optional[int] = None, address_operator: Optional[str] = None, fundsAddr: Optional[str] = None,
-                       expected_assert_code: Optional[int] = None, expected_assert_msg: Optional[str] = None) -> Optional[str]:
-        if (expected_assert_code and not expected_assert_msg) or (not expected_assert_code and expected_assert_msg):
-            raise AssertionError("Intending to use assert_raises_rpc_error() but didn't specify code and message")
+    def update_service(self, node: TestNode, submit: bool, addrs_core_p2p: Union[str, List[str], None] = None, platform_node_id: Optional[str] = None,
+                       addrs_platform_p2p: Union[int, str, List[str], None] = None, addrs_platform_https: Union[int, str, List[str], None] = None,
```
Comment (knst) [inline test/functional/test_framework/test_framework.py:1414]: is there any benefits to have type `Union[int, str, List[str], None]` specified?
It makes code a messy but it seems as almost everything will fit this Union

File: test/functional/rpc_netinfo.py [hunk 26]
```diff
@@ -286,13 +453,108 @@ def test_deprecation(self):
...
+
+        # Restart the client to see if (de)ser works as intended (CDeterministicMNStateDiff is a special case and we just made an update)
+        self.node_evo.set_active_state(self, False)
+        self.reconnect_nodes()
+
+        # Check that 'service' correctly reports as coreP2PAddrs[0]
+        proregtx_rpc = self.node_simple.getrawtransaction(proregtx_hash, True)
+        proupservtx_rpc = self.node_simple.getrawtransaction(proupservtx_hash, True)
+        assert_equal(proregtx_rpc['proRegTx']['service'], f"127.0.0.1:{self.node_evo.mn.nodePort}")
+        assert_equal(proupservtx_rpc['proUpServTx']['service'], f"127.0.0.1:{self.node_evo.mn.nodePort}")
```
Comment (knst) [inline test/functional/rpc_netinfo.py:557]: (out-of-scope, refactoring suggestion)
`rpc_netinfo.py` is really long now! Is there any way to make it more compact? maybe de-duplicate code somehow?

Comment (kwvg) [reply test/functional/rpc_netinfo.py:557]: Due to [dash#6811](https://github.com/dashpay/dash/pull/6811), `rpc_netinfo.py` is expected to reduce in size because the deprecation will be reversed (see 5ddc24d5139f5fe28920ce7b0926efb47b5e67bd)

File: test/functional/test_framework/test_framework.py (hunk 25 above)
Comment (kwvg) [reply test/functional/test_framework/test_framework.py:1414]: The explicit type definition serves two purposes, to allow the linter to apply stricter validation and to document what we expect to accept, `Union[str, None]` is equivalent to `Optional[str]` but `Optional[str, int, List[str]]` is not allowed, hence the more verbose `Union`. The union seems permissive but critically, it _doesn't_ allow `List[int]`, which mirrors the restriction that we do not allow specifying multiple ports, it's either a port or a list of addresses.

File: src/rpc/evo_util.cpp [hunk 27]
```diff
@@ -67,22 +68,104 @@ template void ProcessNetInfoCore(CProRegTx& ptx, const UniValue& input, const bo
...
 template <typename T1>
-void ProcessNetInfoPlatform(T1& ptx, const UniValue& input_p2p, const UniValue& input_http)
+void ProcessNetInfoPlatform(T1& ptx, const UniValue& input_p2p, const UniValue& input_http, const bool optional)
 {
     CHECK_NONFATAL(ptx.netInfo);
 
-    auto process_field = [](uint16_t& target, const UniValue& input, const std::string& field_name) {
-        if (!input.isNum() && !input.isStr()) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid param for %s, must be number", field_name));
+    auto process_field = [&](uint16_t& maybe_target, const UniValue& input, const NetInfoPurpose purpose,
```
Comment (knst) [inline src/rpc/evo_util.cpp:67]: 1. there is a lot of duplicated code in functions `ProcessNetInfoPlatform` and `ProcessNetInfoCore`
2. there is a lot of duplicated error strings
3. names of param and strings that returned by helper PurposeToString are not matched (for example, "platformP2PAddrs"  and "platform_p2p")
4. there's a lot of "return; // nothing  to do" that makes code difficult to read - it's not linear, but a bit messy
5. error strings for similar failures are not always matching, for example "(must specify port number)" an "(must specify port number only)" - I re-write some of them and unified.

Please, check commits in my branch https://github.com/knst/dash/commits/refactor-rpc-util/
Commits here are small as possible and atomic to make it review easier
I prefer if you will pick these commits and squash them to your PR rather than create a new PR because I am not ready to rebase it in case of any major conflict.

File: src/rpc/coinjoin.cpp (hunk 14 above)
Comment (kwvg) [reply src/rpc/coinjoin.cpp]: CoinJoin-specific RPCs will now return `addrs_core_p2p` instead of `addresses`, resolved in latest push.

File: src/evo/deterministicmns.cpp (hunk 15 above)
Comment (kwvg) [reply src/evo/deterministicmns.cpp]: `MainParams()` have been exported from `evo/netinfo.cpp` and used in `evo/deterministicmns.cpp`, resolved in latest push.

File: src/evo/dmnstate.cpp (hunk 19 above)
Comment (kwvg) [reply src/evo/dmnstate.cpp:162]: JSON is order-indifferent, this shouldn't be considered breaking

File: src/evo/core_write.cpp (hunk 23 above)
Comment (kwvg) [reply src/evo/core_write.cpp]: `ShimNetInfoPlatform()` has been renamed to `GetNetInfoWithLegacyFields()` and `ShimPlatformPort()` has been renamed to `GetPlatformPort()`, resolved in latest push.

File: src/evo/netinfo.h (hunk 24 above)
Comment (kwvg) [reply src/evo/netinfo.h:296]: Docs added for `ProcessCandidate()` and `ValidateService()`, resolved in latest push.

File: src/rpc/evo_util.cpp (hunk 27 above)
Comment (kwvg) [reply src/rpc/evo_util.cpp:67]: Most changes have been implemented but the `{core,platform}{P2P,HTTPS}Addrs` naming convention has been retained as they represent the name of the _field_ with respect to the RPC, _not_ how it is internally represented by (e.g. the old name `ipAndPort` doesn't reflect the variable name, `addr`)

**Review by knst — APPROVED** [review]
LGTM 90713a2f2aeea9247bbbfe891922de24e46368d5
